src/orange/OrangeTypes.hh

```diff
@@ -436,6 +424,67 @@ CELER_CONSTEXPR_FUNCTION Sense to_sense(SignedSense s)
     return Sense(static_cast<int>(s) >= 0);
 }
 
+//---------------------------------------------------------------------------//
+/*!
+ * Wrapper for a sense value which associated metadata.
+ */
+class SenseMeta
+{
+    /*!
+     * Bitset indices for the sense metadata.
+     */
+    enum SenseFlags : unsigned char
+    {
+        cache,  //!< Sense is cached
+        sense,  //!< Sense value
+    };
+
+  public:
+    constexpr SenseMeta() = default;
+
+    //! Construct with a sense value
+    CELER_CONSTEXPR_FUNCTION SenseMeta(Sense sense)
+    {
+        sense_[SenseFlags::sense] = static_cast<bool>(sense);
+        sense_[SenseFlags::cache] = true;
+    }
+
+    //! Convert to a sense value
+    CELER_CONSTEXPR_FUNCTION operator Sense() const
+    {
+        return to_sense(sense_[SenseFlags::sense]);
+    }
+
+    //! Convert to a boolean value
+    CELER_CONSTEXPR_FUNCTION explicit operator bool() const
+    {
+        return static_cast<bool>(sense_[SenseFlags::sense]);
+    }
+
+    //! Assign a sense value
+    CELER_CONSTEXPR_FUNCTION SenseMeta& operator=(Sense sense)
+    {
+        sense_[SenseFlags::sense] = static_cast<bool>(sense);
+        sense_[SenseFlags::cache] = true;
+        return *this;
+    }
+
+    //! Check wether there is a cached sense value
+    CELER_CONSTEXPR_FUNCTION bool cached() const
+    {
+        return sense_[SenseFlags::cache];
+    }
+
+    //! Invalidate the cached sense value
+    CELER_CONSTEXPR_FUNCTION void invalidate()
+    {
+        sense_[SenseFlags::cache] = false;
+    }
+
+  private:
+    Bitset<2> sense_;
```

Comment:
Hm, doesn't this increase `sizeof(Sense) == 1` to `sizeof(Sense) == unsigned int`? Should our bitset have an additional "underlying type" parameter, or based on the size of `N` increase the word size up to a certain point?
```c++
using word_type = std::conditional<(N <= 8), std::uint_least8_t,
                  std::conditional<(N <= 16), std::uint_least16_t,
                  unsigned int>>;
```

Reply:
that would be possible, but a bit tricky, we have to be careful because `std::uint8_t` is promoted to a int for `operator ~` which causes issue, maybe do so in a separate PR?